sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java

```diff
@@ -35,6 +36,33 @@
  * Clients should use {@link FileSystems} utility.
  */
 public abstract class FileSystem<ResourceIdT extends ResourceId> {
+  /**
+   * This is the entry point to convert user-provided specs to {@link ResourceIdT ResourceIds}.
+   * Callers should use {@link #match} to resolve users specs ambiguities before
+   * calling other methods.
+   *
+   * <p>Implementation should handle the following ambiguities of a user-provided spec:
+   * <ul>
+   * <li>1). spec could be a glob or a uri. {@link #match} should be able to tell and
+   * choose efficient implementations.
+   * <li>2). spec does not end with a path delimiter may refer to files or directories:
+   * For example, local directory "/home/dir/" should be returned for spec “/home/dir”.
+   * (However, spec ends with a path delimiter always refers to directories.)
```

Comment:
How about something like, `The user-provided {@code spec} might refer to a file or a directory. It is common that users that wish to indicate a directory will omit the trailing {@code /}, such as in a spec of {@code "/tmp/dir"}. The {@link FileSystem} should be able to recognize a directory with the trailing {@code /} omitted, but should always return a correct {@link ResourceId} (e.g., {@code "/tmp/dir/"} inside the returned {@link MatchResult}.`